### PR TITLE
Validate crop specification against image

### PR DESCRIPTION
**Pull request by kenoir**
Cropper can be asked to perform impossible crops which can cause it to
behave strangely (500s thrown from im4java as crop fails, endless waits to perform impossibly huge crops).

This changeset validates crop requests against the source image
dimensions before performing a crop. Invalid crops will result in a HTTP
400 Bad Request.

### CODE REVIEW

**Comment by kenoir** [comment]
This should prevent some of the Cropper 500s occurring.

File: cropper/app/lib/Crops.scala [hunk 1]
```diff
@@ -52,11 +53,25 @@ object Crops {
     else
       Config.landscapeCropSizingWidths.filter(_ <= bounds.width).map(w => Dimensions(w, math.round(w / aspectRatio)))
 
+  def isInvalidCrop(asset: Asset, specification: CropSource): Boolean = {
+    asset.dimensions.map { dimensions =>
+      val bounds = specification.bounds;
```
Comment (theefer) [inline cropper/app/lib/Crops.scala]: Semi-colons are for Java.

File: cropper/app/controllers/Application.scala [hunk 2]
```diff
@@ -86,6 +86,7 @@ object Application extends Controller with ArgoHelpers {
         } recover {
           case InvalidImage => respondError(BadRequest, "invalid-image", InvalidImage.getMessage)
           case MissingSecureSourceUrl => respondError(BadRequest, "no-source-image", MissingSecureSourceUrl.getMessage)
+          case InvalidCropRequest => respondError(BadRequest, "no-source-image", InvalidCropRequest.getMessage)
```
Comment (theefer) [inline cropper/app/controllers/Application.scala]: Wrong error key?

**Comment by theefer** [comment]
Minor comments, nice otherwise! :+1:

**Comment by theefer** [comment]
Was this related to https://github.com/guardian/grid/issues/755 ?

**Comment by kenoir** [comment]
Fixes #755

**Pull request merged**